cmake/DARTMacros.cmake

```diff
@@ -170,3 +170,82 @@ function(dart_format_add)
     endif()
   endforeach()
 endfunction()
+
+#===============================================================================
+# Macro to check for visibility capability in compiler
+macro(check_compiler_visibility variable)
+  include(CheckCXXCompilerFlag)
+  check_cxx_compiler_flag(-fvisibility=hidden ${variable})
+endmacro()
+
+#===============================================================================
+# Function to create an export header for control of binary symbols visibility.
+#
+# dart_add_export_file(<target_name>
+#     [BASE_NAME <base_name>]
+#     [COMPONENT_PATH <component_path>]
+function(dart_add_export_file target_name)
+  # Parse boolean, unary, and list arguments from input.
+  # Unparsed arguments can be found in variable ARG_UNPARSED_ARGUMENTS.
+  set(prefix _geh)
+  set(options "")
+  set(oneValueArgs BASE_NAME COMPONENT_PATH)
+  set(multiValueArgs "")
+  cmake_parse_arguments(
+    "${prefix}" "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN}
+  )
+
+  include(GNUInstallDirs)
+
+  # Base name
+  set(base_name "${target_name}")
+  string(REPLACE "-" "_" base_name ${base_name})
+  if(_geh_BASE_NAME)
+    set(base_name ${_geh_BASE_NAME})
+  endif()
+  string(TOUPPER ${base_name} base_name_upper)
+  string(TOLOWER ${base_name} base_name_lower)
+
+  # Componenet path
+  set(component_path "dart")
+  if(_geh_COMPONENT_PATH)
+    set(component_path "dart/${_geh_COMPONENT_PATH}")
+  endif()
+  set(component_detail_export_path "${component_path}/detail/export.h")
+
+  # Generate CMake's default export header
+  include(GenerateExportHeader)
+  generate_export_header(${target_name}
+    EXPORT_MACRO_NAME ${base_name_upper}_DETAIL_API
+    EXPORT_FILE_NAME detail/export.h
+    DEPRECATED_MACRO_NAME ${base_name_upper}_DETAIL_DEPRECATED
+  )
+  install(FILES ${CMAKE_CURRENT_BINARY_DIR}/detail/export.h
+    DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/${component_path}/detail
+    COMPONENT headers
+  )
+
+  # Generate final export header
+  set(header_guard_name ${base_name_upper}_EXPORT_HPP_)
```

Comment:
This would probably be much cleaner if we created a `export.hh.in` and then called `configure_file` on it.

Reply:
Agree. I wrote this lines because I thought each component requires individual `export.hpp.in`, but this is not true.

Reply:
Agree. I wrote this lines because I thought each component requires individual `export.hpp.in`, but this is not true.